nltk/text.py

```diff
@@ -322,18 +341,37 @@ def __len__(self):
     # Interactive console methods
     #////////////////////////////////////////////////////////////
 
-    def concordance(self, word, width=79, lines=25):
-        """
-        Print a concordance for ``word`` with the specified context window.
+    def concordance(self, word, width=79, lines=None, stdout=True):
+        """Generate a concordance for ``word`` with the specified context window.
         Word matching is not case-sensitive.
+
+        :param word: The target word
+        :type word: str
+        :param width: The width of each line, in characters (default=80)
+        :type width: int
+        :param lines: The number of lines to display (default=25)
+-       :type lines: int
+        :param stdout: print concordance [True] or return list [False]
+        :type stdout: bool
+
+        :return: concordance, either printed to stdout or as a list
+        :rtype: None, list
+
+        :param stdout: print concordance to stdout
+        :type stdout: bool
+
         :seealso: ``ConcordanceIndex``
         """
         if '_concordance_index' not in self.__dict__:
-            #print("Building index...")
+            # print("Building index...")
             self._concordance_index = ConcordanceIndex(self.tokens,
                                                        key=lambda s:s.lower())
+        self._concordance_index.generate_concordance(word, width)
 
-        self._concordance_index.print_concordance(word, width, lines)
+        if stdout:
+            self._concordance_index.print_concordance(lines)
+        else:
+            return self._concordance_index.list_concordance(lines)
```

Comment:
L365-L374 shows the effect of the spaghetti code where lots if-else is required to patch how we control what is generated and what is printed. 

Let's try and work out the better way of adding a generate vs print functionality as a parameter to the function than keeping multiple functions in the codebase.